### PR TITLE
DEV: Use service objects for chat thread lookup

### PR DESCRIPTION
**Pull request by martin-brennan**
Adds a new LookupThread class that handles finding the
thread based on thread + channel ID, checking permissions
and policy/contract checks.

### CODE REVIEW

**Comment by jjaffeux** [comment]
@loic @martin will write a spec for this tomorrow, but if you already have comments, feel free to add them here

File: plugins/chat/app/controllers/api/chat_channel_threads_controller.rb
Comment (Flink) [inline plugins/chat/app/controllers/api/chat_channel_threads_controller.rb:5]: We can remove this actually, as we’re checking input parameters directly in the service through its contract.

File: plugins/chat/app/services/lookup_thread.rb
Comment (Flink) [inline plugins/chat/app/services/lookup_thread.rb:31]: This doesn’t work actually :sweat_smile: (I’ve just checked). We should rewrite the contract as:
```ruby
attribute :thread_id, :integer
attribute :channel_id, :integer

validates :thread_id, :channel_id, presence: true
```

File: plugins/chat/app/services/lookup_thread.rb
Comment (Flink) [inline plugins/chat/app/services/lookup_thread.rb:41]: As we’re validating parameters through the contract, we should probably take values from it (even if here it won’t have a big impact):
```ruby
def fetch_thread(contract:, **)
  …
  .find_by(id: contract.thread_id, channel_id: contract.channel_id)
end
```

File: plugins/chat/app/services/lookup_thread.rb
Comment (Flink) [inline plugins/chat/app/services/lookup_thread.rb:45]: We don’t need to repeat `includes` each time, this works too:
```ruby
        ChatThread.includes(
          :channel,
          original_message_user: :user_status,
          original_message: :chat_webhook_event,
        )
```

File: plugins/chat/app/services/lookup_thread.rb
Comment (Flink) [inline plugins/chat/app/services/lookup_thread.rb:46]: We should use `find_by` instead of `find_by!` as the latter will raise an exception that won’t be caught by the service. While this is working in the controller calling this service because we have a `rescue_from ActiveRecord::RecordNotFound` defined somewhere, if we were to call the service in another context, the exception will have to be handled manually. (instead of just being able to inspect the result object)
We’ll have to add this line of code in the controller though:
```ruby
      on_model_not_found(:thread) { raise Discourse::NotFound }
```
This way, everything is “self-contained” inside the service.

**Review by Flink — COMMENTED** [review]
Some suggestions to improve things a bit, but very nice PR :blush:

File: plugins/chat/app/services/lookup_thread.rb
Comment (martin-brennan) [reply plugins/chat/app/services/lookup_thread.rb:46]: I will change this one, but I do wonder why we don't just handle `ActiveRecord::NotFound` inside the `ModelStep` class, same as how we rescue `ArgumentError`? Then whoever is writing the service doesn't need to worry about it.

https://github.com/discourse/discourse/blob/1eaebd686439e0a2b32bbbd5f07fc7e32f3efca6/plugins/chat/app/services/base.rb#L216-L225

File: plugins/chat/app/services/lookup_thread.rb
Comment (martin-brennan) [reply plugins/chat/app/services/lookup_thread.rb]: I changed the docs slightly on Base to indicate this is the best practice

File: plugins/chat/app/services/lookup_thread.rb
Comment (martin-brennan) [reply plugins/chat/app/services/lookup_thread.rb]: Also why raise `ArgumentError` here instead of just `Discourse::NotFound`?

**Comment by martin-brennan** [comment]
Thanks for the review @Flink , I've addressed comments and added the spec now, and added some other comments too :+1:

File: plugins/chat/spec/support/chat_service_matcher.rb
Comment (martin-brennan) [inline plugins/chat/spec/support/chat_service_matcher.rb:74]: I removed this because it would error in the scenario I have with `LookupThread`, where this contract step would not exist if the initial `threaded_discussions_enabled` policy does not pass, so my test for "when params are not valid" would just fail with an error.

At any rate I don't see why this `failure_message` is needed anyway, the `FailStep` one fits better IMO

File: plugins/chat/app/services/lookup_thread.rb
Comment (Flink) [reply plugins/chat/app/services/lookup_thread.rb]: The idea is that a service doesn't rely on exceptions for control flow and if an exception were to be raised, then it will interrupt the sequence of steps leaving the service in an unknown/unstable state. And this is good for catching bugs  :grin: 
I remember @jjaffeux having the same thought as you about AR exceptions but dropped it in the end. He will probably explain this better than me.

Here, the `ArgumentError` handling is something we did to prevent the service from breaking when a parameter was missing. As calling the `model` step first without checking for parameters validity before is something that seemed totally fine to do (as in the end we’re just passing one argument to `#find_by` typically), it wouldn’t have been nice to have the whole service break just for a missing argument. Maybe that’s something we should revisit though. We could decide to remove this `ArgumentError` logic altogether and to rely on a contract instead. We could also simply decide that if an expected parameter isn’t present, then we should treat that as a bug and not as something expected. (typically when the `id` comes from a Rails route and not from a query arg, we can treat it as a dependency instead of an input parameter)

File: plugins/chat/spec/support/chat_service_matcher.rb
Comment (Flink) [reply plugins/chat/spec/support/chat_service_matcher.rb:74]: Indeed, that’s better with the default `failure_message`. I think those are just leftovers and we can actually refactor these matchers a bit now that we normalized everything regarding the result keys.

I’m going to open a small refactor PR :+1:

File: plugins/chat/spec/services/lookup_thread_spec.rb
Comment (Flink) [inline plugins/chat/spec/services/lookup_thread_spec.rb:25]: Why not define default `fab!` and `params` at the top of the `describe` block and then just change what’s needed for a given example?
This way, when looking at the example, it’s a bit easier to understand what differs from a previous example.

**Comment by jjaffeux** [comment]
@martin-brennan I asked @Flink to push the changes he requested in second round of feedback. OFC we can discuss this more in chat-team channel, or you can just open another PR on specific points if you disagree. I just want to move on here and avoid too many back and forth while you are focused on other thread subjects.

File: plugins/chat/spec/services/lookup_thread_spec.rb
Comment (martin-brennan) [reply plugins/chat/spec/services/lookup_thread_spec.rb:31]: Oh interesting, will fix

File: plugins/chat/spec/services/lookup_thread_spec.rb
Comment (martin-brennan) [reply plugins/chat/spec/services/lookup_thread_spec.rb:25]: This kind of goes against the style @jjaffeux has been doing -- I used to do it the other way as well. We should decide on which way we want to go and stick to it.

File: plugins/chat/spec/services/lookup_thread_spec.rb
Comment (martin-brennan) [reply plugins/chat/spec/services/lookup_thread_spec.rb:16]: will fix 👍

File: plugins/chat/spec/support/chat_service_matcher.rb
Comment (martin-brennan) [reply plugins/chat/spec/support/chat_service_matcher.rb:74]: Cool sounds good 👍

File: plugins/chat/app/services/lookup_thread.rb
Comment (martin-brennan) [reply plugins/chat/app/services/lookup_thread.rb]: Yep thanks for the explanation. I do remember the control flow discussions, I mostly meant why `raise ArgumentError` here instead of `raise Discourse::NotFound`. All good I see the intent now.

**Comment by martin-brennan** [comment]
@Flink pushed review changes again :) I think we can approve/merge this now, once we sort out the last comment around spec styles

File: plugins/chat/spec/services/lookup_thread_spec.rb
Comment (jjaffeux) [reply plugins/chat/spec/services/lookup_thread_spec.rb:28]: Im fine trying the let way

**Comment by jjaffeux** [comment]
Looks great, let's merge this 👌

**Review by jjaffeux — APPROVED** [review]

**Pull request merged**
